airflow/models.py

```diff
@@ -119,12 +119,17 @@ def clear_task_instances(tis, session, activate_dag_runs=True):
             if ti.job_id:
                 ti.state = State.SHUTDOWN
                 job_ids.append(ti.job_id)
-        # todo: this creates an issue with the webui tests
-        # elif ti.state != State.REMOVED:
-        #     ti.state = State.NONE
-        #     session.merge(ti)
         else:
-            session.delete(ti)
+            try:
+                ti_retries = dag.get_task(ti.task_id).retries
+                ti.max_tries = ti.try_number + ti_retries
+            except (AttributeError, AirflowException):
+                # When dag is None or task not found in dag, we
+                # keep max_tries the maximum value of its original
+                # max_tries or the current task try number
+                ti.max_tries = max(ti.max_tries, ti.try_number)
```

Comment:
@bolkedebruin max_tries is maximum number of tries a task can have, i.e when a task fails, we check try_number < max_tries to see it we should change the status to UP_FOR_RETRY. It is set to `task.retries` at first.